### PR TITLE
Detect current network and netmask

### LINKED ISSUE #13211: The openvpn chart doesn't correctly determine the subnet mask for one of it's routes.
Chart: stable/openvpn
Type: bug
Priority: low?

It appears [this line](https://github.com/helm/charts/blob/856302e6cdd0f763ce105722076140afa5588320/stable/openvpn/templates/config-openvpn.yaml#L104) makes an assumption on what the subnet mask should be and in our case the wrong assumption. which means we get the following error client side: 
```text
Error: Invalid prefix for given prefix length.
Tue Apr 23 17:32:59 2019 ERROR: Linux route add command failed: external program exited with error status: 2
```
As in our case it is a subset of the subnet we supply using `OVPN_K8S_POD_NETWORK` in our values file it is only a cosmetic issue for us. There are likely circumstances where this isn't the case however. Perhaps the subnet mask can be retrieved at the same time as the network. Also simply replacing the last octet with a 0 could cause problems when people use smaller subnets then /24.

### PR DESCRIPTION
**Pull request by baelish**
@jfelten Please consider this pull request.

#### What this PR does / why we need it:
This PR detects the IP address and Netmask correctly. Which results in only valid routes being passed to the clients.

Previously the network was guessed (assumed it was a /24 by replacing the last octet with a 0) and then later the netmask was hardcoded to 255.255.240.0 as a result the requested route was almost always incorrect. As most people supply their own or use the vpn tunnel as a default gateway it was probably rare to cause any issues.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #13211

#### Special notes for your reviewer:
the `cidr2mask()` function was copied from [here](https://forum.archive.openwrt.org/viewtopic.php?id=47986&p=1#p220781) the rest was written from scratch.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

File: stable/openvpn/templates/config-openvpn.yaml [hunk 1]
```diff
@@ -65,9 +94,16 @@ data:
...
       fi
-      IP=$(ip route get 8.8.8.8 | awk '/8.8.8.8/ {print $NF}')
-      BASEIP=`echo $IP | cut -d"." -f1-3`
-      NETWORK=`echo $BASEIP".0"`
+
+      intAndIP="$(ip route get 8.8.8.8 | awk '/8.8.8.8/ {print $5 "-" $7}')"
+      int="${intAndIP%-*}"
+      ip="${intAndIP#*-}"
+
+      echo $int "and" $ip
```
Comment (jasongwartz) [inline stable/openvpn/templates/config-openvpn.yaml]: intentional or leftover debug?

File: stable/openvpn/templates/config-openvpn.yaml [hunk 2]
```diff
@@ -78,6 +114,7 @@ data:
       sed 's|OVPN_K8S_SEARCH|'"${FORMATTED_SEARCH}"'|' -i /etc/openvpn/openvpn.conf
       sed 's|OVPN_K8S_DNS|'"${DNS}"'|' -i /etc/openvpn/openvpn.conf
       sed 's|NETWORK|'"${NETWORK}"'|' -i /etc/openvpn/openvpn.conf
+      sed 's|NETMASK|'"${NETMASK}"'|' -i /etc/openvpn/openvpn.conf
```
Comment (jasongwartz) [inline stable/openvpn/templates/config-openvpn.yaml:114]: maybe I'm missing something here, but I'm guessing we want to put the NETMASK in the openvpn.conf file somewhere? I don't see `NETMASK` in the openvpn.conf file as a placeholder at the moment - are you using `.Values.openvpn.OVPN_SUBNET` in its place?

File: stable/openvpn/templates/config-openvpn.yaml [hunk 3]
```diff
@@ -53,6 +53,35 @@ data:
...
+          ip="${1%/*}"
+          mask="${1#*/}"
+          octets=$(echo "$ip" | tr '.' '\n')
+
+          for octet in $octets; do
+              i=$((i+1))
+              if [ $i -le $(( mask / 8)) ]; then
+                  netOctets="$netOctets.$octet"
+              elif [ $i -eq  $(( mask / 8 +1 )) ]; then
+                  echo $octet
```
Comment (jasongwartz) [inline stable/openvpn/templates/config-openvpn.yaml]: when I run `cidr2net()` locally, my output is:
```
63
10.0.3.0
```
which makes sense, given that my IP is 10.0.3.63. However, I suspect this `echo $octet` might be leftover debug?

**Review by jasongwartz — COMMENTED** [review]
Thanks for the contribution!

This isn't a situation I personally encounter with this chart, so I have a few general questions but in general am just comparing what you've implemented here with the openvpn docs.
Cheers!

File: stable/openvpn/templates/config-openvpn.yaml (hunk 3 above)
Comment (baelish) [reply stable/openvpn/templates/config-openvpn.yaml]: 130fbbb Leftover debug.

File: stable/openvpn/templates/config-openvpn.yaml (hunk 1 above)
Comment (baelish) [reply stable/openvpn/templates/config-openvpn.yaml]: 130fbbb leftover debug

File: stable/openvpn/templates/config-openvpn.yaml (hunk 2 above)
Comment (baelish) [reply stable/openvpn/templates/config-openvpn.yaml:114]: db363d6 Well spotted, embarrasingly my testing didn't catch that as the helm upgrade I performed didn't rewrite the openvpn.conf file which I had manually edited to contain the correction in the mentioned commit.

**Comment by baelish** [comment]
> Thanks for the contribution!
> 
> This isn't a situation I personally encounter with this chart, so I have a few general questions but in general am just comparing what you've implemented here with the openvpn docs.
> Cheers!

You're welcome. It's just a cosmetic issue in most cases I'm sure. Thanks for the review, I believe I've addressed the things you've mentioned and my, now more thorough, testing shows things are working as I would expect them to.

Cheers!

**Comment by jasongwartz** [comment]
thanks! one request - can you test that this change still works with the kylemanna/openvpn docker image, as per the change coming in https://github.com/helm/charts/pull/12269 ?

**Comment by baelish** [comment]
> thanks! one request - can you test that this change still works with the kylemanna/openvpn docker image, as per the change coming in #12269 ?

No worries. 

TL;DR: I expect there to be no issues related to the changes in this pull request when the image is changed. 

I tested by simply updating the image in the current chart, this failed because of easyrsa not available in the right location: 

```text
~/dev/charts/stable/openvpn issue-13211
thymara$>kubectl logs --namespace=hd-test testy-tiger-openvpn-75b87d7854-7w5mm
/etc/openvpn/setup/setup-certs.sh: line 10: ./easyrsa: No such file or directory
/etc/openvpn/setup/setup-certs.sh: line 11: ./easyrsa: No such file or directory
/etc/openvpn/setup/setup-certs.sh: line 12: ./easyrsa: No such file or directory
/etc/openvpn/setup/setup-certs.sh: line 13: ./easyrsa: No such file or directory
/etc/openvpn/setup/configure.sh: 37: [: unexpected operator
Options error: --dh fails with '/etc/openvpn/certs/pki/dh.pem': No such file or directory
Options error: --ca fails with '/etc/openvpn/certs/pki/ca.crt': No such file or directory
Options error: --cert fails with '/etc/openvpn/certs/pki/issued/server.crt': No such file or directory
Options error: --key fails with '/etc/openvpn/certs/pki/private/server.key': No such file or directory
Options error: Please correct these errors.
Use --help for more information.
```
As this is unrelated to the changes we are discussing I just tested on the docker image directly. This looks fine:

```text
~/dev/charts/stable/openvpn issue-13211
thymara$>docker run -ti kylemanna/openvpn sh
/ #       cidr2mask() {
>          # Number of args to shift, 255..255, first non-255 byte, zeroes
>          set -- $(( 5 - ($1 / 8) )) 255 255 255 255 $(( (255 << (8 - ($1 % 8))) & 255 )) 0 0 0
>          [ $1 -gt 1 ] && shift "$1" || shift
>          echo ${1-0}.${2-0}.${3-0}.${4-0}
>       }
/ # 
/ #       cidr2net() {
>           local i ip mask netOctets octets
>           ip="${1%/*}"
>           mask="${1#*/}"
>           octets=$(echo "$ip" | tr '.' '\n')
> 
>           for octet in $octets; do
>               i=$((i+1))
>               if [ $i -le $(( mask / 8)) ]; then
>                   netOctets="$netOctets.$octet"
>               elif [ $i -eq  $(( mask / 8 +1 )) ]; then
>                   netOctets="$netOctets.$((((octet / ((256 / ((2**((mask % 8)))))))) * ((256 / ((2**((mask % 8))))))))"
>               else
>                   netOctets="$netOctets.0"
>               fi
>           done
> 
>           echo ${netOctets#.}
>       }
/ #       intAndIP="$(ip route get 8.8.8.8 | awk '/8.8.8.8/ {print $5 "-" $7}')"
/ #       int="${intAndIP%-*}"
/ #       ip="${intAndIP#*-}"
/ #       cidr="$(ip addr show dev "$int" | awk -vip="$ip" '($2 ~ ip) {print $2}')"
/ # 
/ #       NETWORK="$(cidr2net $cidr)"
/ #       NETMASK="$(cidr2mask ${cidr#*/})"
/ # echo $NETWORK
172.17.0.0
/ # echo $NETMASK
255.255.0.0
/ # ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
109: eth0@if110: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.17.0.3/16 brd 172.17.255.255 scope global eth0
       valid_lft forever preferred_lft forever
/ # 
```

**Comment by jasongwartz** [comment]
@baelish sounds good to me, thanks for checking!
